internal/engine/experiment/echcheck/utls.go

```diff
@@ -0,0 +1,89 @@
+package echcheck
+
+import (
+	"context"
+	"crypto/tls"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	utls "gitlab.com/yawning/utls.git"
+	"net"
+)
+
+type tlsHandshakerWithExtensions struct {
+	conn       utlsConn
+	extensions []utls.TLSExtension
+	dl         model.DebugLogger
+	id         *utls.ClientHelloID
+}
+
+var _ model.TLSHandshaker = &tlsHandshakerWithExtensions{}
+
+// newHandshakerWithExtensions returns a NewHandsharer function for creating
+// tlsHandshakerWithExtensions instances.
+func newHandshakerWithExtensions(extensions []utls.TLSExtension) func(dl model.DebugLogger, id *utls.ClientHelloID) model.TLSHandshaker {
+	return func(dl model.DebugLogger, id *utls.ClientHelloID) model.TLSHandshaker {
+		return &tlsHandshakerWithExtensions{
+			extensions: extensions,
+			dl:         dl,
+			id:         id,
+		}
+	}
+}
+
+func (t *tlsHandshakerWithExtensions) Handshake(ctx context.Context, conn net.Conn, tlsConfig *tls.Config) (
+	net.Conn, tls.ConnectionState, error) {
+	t.conn = newConnUTLSWithHelloID(conn, tlsConfig, t.id)
+
+	if t.extensions != nil && len(t.extensions) != 0 {
+		t.conn.BuildHandshakeState()
+		t.conn.Extensions = append(t.conn.Extensions, t.extensions...)
+	}
+
+	err := t.conn.Handshake()
+
+	return t.conn.NetConn(), t.conn.ConnectionState(), err
+}
+
+// utlsConn is a utls connection
+type utlsConn struct {
```

Comment:
I noticed that you had to re-implement `utlsConn` because it is private in `netxlite`. This made us realize that maybe we should make the relevant types and functions in `netxlite/utls.go` public in order to simplify the use of utls in experiments.

Reply:
I've added a TODO to remove this once https://github.com/ooni/probe/issues/2378 is resolved. I will try to submit a separate PR for that.